File: java/ankit_kanyal/devnest_problems/PartitionList.java
// link- https://devsnest.in/algo-challenges/partition-list?tab=question

static Node solve(Node head, int x){
    if(head==null || head.next==null) return head;

    Node small=null,smaller=null;
    Node big=null,bigger=null;

    while(head!=null){
        if(head.data<x){
          if(small==null){
              small=head;
              smaller=head;
          }else{
             small.next=head;
             small=head;
          }
        }else{
           if(big==null){
              big=head;
              bigger=head;
           }else{
             big.next=head;
             big=head;
           }
        }
        head=head.next;
    } 

    if(small==null) return bigger;
    if(big==null) return smaller;

    small.next=bigger;
    big.next=null;

    return smaller;
}

File: Python/chandu/Partition_List.py
def solve(head, x):
    # CODE HERE
    l1 = l1_head = Node(-1)
    l2 = l2_head = Node(-1)

    while head:
        if head.data <x:
            l1.next = head
            l1 = l1.next
        
        else:
            l2.next = head
            l2 = l2.next
        
        head = head.next
    
    # l2 should end to None
    l2.next = None
    # combine
    l1.next = l2_head.next

    return l1_head.next
